Split package problem when using java modules

langchain4j can not be used in a Java project that is working on the module path. By doing so the following error happens:

`error: the unnamed module reads package dev.langchain4j.retriever from both langchain4j and langchain4j.core
`
That is based on having the same package `dev.langchain4j.retriever` (with content) in both modules `langchain4j` and `langchain4j.core`.

![image](https://github.com/langchain4j/langchain4j/assets/9443847/0337d42e-d0cf-4643-86ce-5b70d42e51e8)

A project like langchain4j should be 100% compatible with the Java module system since that is used by more and more libs over time.

I did a lot of work in that area and happy to help if you want to work on that topic.

@langchain4j  To do full support for the module system, we need to add at least basic names to each module. While for full support, a `module-info.java` would be needed, the easiest way is to add the info to the `MANIFEST` of the jar as an `automatic-module-name`. You can find more information about that here https://dev.java/learn/modules/automatic-module/
Based on that, I created a branch and defined the names of each module. You can find my changes in this draft: https://github.com/langchain4j/langchain4j/pull/1502

Sadly I can not run `mvn install` on my machine to verify that everything works for a project that uses modules since the build on my machine always ends with the following errors. I did the same for main branch to verify that it is not my change that did it. All errors happen for me on mainbranch, too.

```
[INFO] --- compiler:3.12.1:testCompile (default-testCompile) @ langchain4j ---
[INFO] Recompiling the module because of changed dependency.
[INFO] Compiling 37 source files with javac [debug target 8] to target/test-classes
[INFO] -------------------------------------------------------------
[ERROR] COMPILATION ERROR : 
[INFO] -------------------------------------------------------------
[ERROR] /Users/hendrikebbers/git/langchain4j/langchain4j/src/test/java/dev/langchain4j/service/AiServicesWithRagIT.java:[12,39] cannot find symbol
  symbol:   class AllMiniLmL6V2QuantizedEmbeddingModel
  location: package dev.langchain4j.model.embedding
[ERROR] /Users/hendrikebbers/git/langchain4j/langchain4j/src/test/java/dev/langchain4j/store/embedding/inmemory/InMemoryEmbeddingStoreSerializedTest.java:[4,39] cannot find symbol
  symbol:   class AllMiniLmL6V2QuantizedEmbeddingModel
  location: package dev.langchain4j.model.embedding
[ERROR] /Users/hendrikebbers/git/langchain4j/langchain4j/src/test/java/dev/langchain4j/classification/EmbeddingModelTextClassifierTest.java:[3,39] cannot find symbol
  symbol:   class AllMiniLmL6V2QuantizedEmbeddingModel
  location: package dev.langchain4j.model.embedding
[ERROR] /Users/hendrikebbers/git/langchain4j/langchain4j/src/test/java/dev/langchain4j/store/embedding/inmemory/InMemoryEmbeddingStoreRemovalTest.java:[4,39] cannot find symbol
  symbol:   class AllMiniLmL6V2QuantizedEmbeddingModel
  location: package dev.langchain4j.model.embedding
[ERROR] /Users/hendrikebbers/git/langchain4j/langchain4j/src/test/java/dev/langchain4j/store/embedding/inmemory/InMemoryEmbeddingStoreTest.java:[6,39] cannot find symbol
  symbol:   class AllMiniLmL6V2QuantizedEmbeddingModel
  location: package dev.langchain4j.model.embedding
[INFO] 5 errors 
```

If I ignore test code and do a `mvn compile` I can go some steps further but end with the following errors:

```
[INFO] --- compiler:3.12.1:compile (default-compile) @ langchain4j-easy-rag ---
[INFO] Recompiling the module because of changed dependency.
[INFO] Compiling 1 source file with javac [debug target 8] to target/classes
[INFO] -------------------------------------------------------------
[ERROR] COMPILATION ERROR : 
[INFO] -------------------------------------------------------------
[ERROR] /Users/hendrikebbers/git/langchain4j/langchain4j-easy-rag/src/main/java/dev/langchain4j/data/document/splitter/recursive/RecursiveDocumentSplitterFactory.java:[5,39] cannot find symbol
  symbol:   class HuggingFaceTokenizer
  location: package dev.langchain4j.model.embedding
[ERROR] /Users/hendrikebbers/git/langchain4j/langchain4j-easy-rag/src/main/java/dev/langchain4j/data/document/splitter/recursive/RecursiveDocumentSplitterFactory.java:[12,57] cannot find symbol
  symbol:   class HuggingFaceTokenizer
  location: class dev.langchain4j.data.document.splitter.recursive.RecursiveDocumentSplitterFactory
[INFO] 2 errors 
```

--- a/langchain4j/src/main/java/dev/langchain4j/service/AiServiceTokenStream.java
+++ b/langchain4j/src/main/java/dev/langchain4j/service/AiServiceTokenStream.java
@@ -3,7 +3,6 @@ package dev.langchain4j.service;
 import dev.langchain4j.agent.tool.ToolSpecification;
 import dev.langchain4j.data.message.AiMessage;
 import dev.langchain4j.data.message.ChatMessage;
-import dev.langchain4j.exception.IllegalConfigurationException;
 import dev.langchain4j.model.chat.request.ChatRequest;
 import dev.langchain4j.model.chat.response.ChatResponse;
 import dev.langchain4j.model.chat.response.StreamingChatResponseHandler;

--- a/langchain4j/src/main/java/dev/langchain4j/service/AiServices.java
+++ b/langchain4j/src/main/java/dev/langchain4j/service/AiServices.java
@@ -37,7 +37,7 @@ import java.util.concurrent.Future;
 import java.util.function.Function;
 
 import static dev.langchain4j.agent.tool.ToolSpecifications.toolSpecificationFrom;
-import static dev.langchain4j.exception.IllegalConfigurationException.illegalConfiguration;
+import static dev.langchain4j.service.IllegalConfigurationException.illegalConfiguration;
 import static dev.langchain4j.internal.ValidationUtils.ensureNotNull;
 import static dev.langchain4j.spi.ServiceHelper.loadFactories;
 import static java.util.Arrays.asList;

--- a/langchain4j/src/main/java/dev/langchain4j/service/DefaultAiServices.java
+++ b/langchain4j/src/main/java/dev/langchain4j/service/DefaultAiServices.java
@@ -1,6 +1,6 @@
 package dev.langchain4j.service;
 
-import static dev.langchain4j.exception.IllegalConfigurationException.illegalConfiguration;
+import static dev.langchain4j.service.IllegalConfigurationException.illegalConfiguration;
 import static dev.langchain4j.internal.Exceptions.illegalArgument;
 import static dev.langchain4j.internal.Exceptions.runtime;
 import static dev.langchain4j.internal.Utils.isNotNullOrBlank;

--- a/langchain4j/src/main/java/dev/langchain4j/service/IllegalConfigurationException.java
+++ b/langchain4j/src/main/java/dev/langchain4j/service/IllegalConfigurationException.java
@@ -1,4 +1,4 @@
-package dev.langchain4j.exception;
+package dev.langchain4j.service;
 
 import static java.lang.String.format;
 

--- a/langchain4j/src/main/java/dev/langchain4j/service/SystemMessage.java
+++ b/langchain4j/src/main/java/dev/langchain4j/service/SystemMessage.java
@@ -1,7 +1,5 @@
 package dev.langchain4j.service;
 
-import dev.langchain4j.exception.IllegalConfigurationException;
-
 import java.lang.annotation.Retention;
 import java.lang.annotation.Target;
 import java.util.function.Function;

--- a/langchain4j/src/main/java/dev/langchain4j/service/UserMessage.java
+++ b/langchain4j/src/main/java/dev/langchain4j/service/UserMessage.java
@@ -1,7 +1,5 @@
 package dev.langchain4j.service;
 
-import dev.langchain4j.exception.IllegalConfigurationException;
-
 import java.lang.annotation.Retention;
 import java.lang.annotation.Target;
 

--- a/langchain4j/src/main/java/dev/langchain4j/service/output/JsonSchemas.java
+++ b/langchain4j/src/main/java/dev/langchain4j/service/output/JsonSchemas.java
@@ -12,7 +12,7 @@ import java.lang.reflect.Type;
 import java.util.LinkedHashMap;
 import java.util.Optional;
 
-import static dev.langchain4j.exception.IllegalConfigurationException.illegalConfiguration;
+import static dev.langchain4j.service.IllegalConfigurationException.illegalConfiguration;
 import static dev.langchain4j.model.chat.request.json.JsonSchemaElementHelper.jsonObjectOrReferenceSchemaFrom;
 import static dev.langchain4j.service.TypeUtils.getRawClass;
 import static dev.langchain4j.service.TypeUtils.resolveFirstGenericParameterClass;

--- a/langchain4j/src/main/java/dev/langchain4j/service/output/ServiceOutputParser.java
+++ b/langchain4j/src/main/java/dev/langchain4j/service/output/ServiceOutputParser.java
@@ -20,7 +20,7 @@ import java.util.Set;
 import java.util.regex.Matcher;
 import java.util.regex.Pattern;
 
-import static dev.langchain4j.exception.IllegalConfigurationException.illegalConfiguration;
+import static dev.langchain4j.service.IllegalConfigurationException.illegalConfiguration;
 import static dev.langchain4j.internal.ValidationUtils.ensureNotNull;
 import static dev.langchain4j.service.TypeUtils.getRawClass;
 import static dev.langchain4j.service.TypeUtils.resolveFirstGenericParameterClass;

--- a/langchain4j/src/test/java/dev/langchain4j/exception/IllegalConfigurationExceptionTest.java
+++ b/langchain4j/src/test/java/dev/langchain4j/exception/IllegalConfigurationExceptionTest.java
@@ -1,5 +1,6 @@
 package dev.langchain4j.exception;
 
+import dev.langchain4j.service.IllegalConfigurationException;
 import org.assertj.core.api.WithAssertions;
 import org.junit.jupiter.api.Test;
 

--- a/langchain4j/src/test/java/dev/langchain4j/service/AiServiceChatMemoryConfigTest.java
+++ b/langchain4j/src/test/java/dev/langchain4j/service/AiServiceChatMemoryConfigTest.java
@@ -6,7 +6,6 @@ import static org.assertj.core.api.Assertions.assertThat;
 import static org.assertj.core.api.Assertions.assertThatThrownBy;
 import static org.mockito.Mockito.verify;
 
-import dev.langchain4j.exception.IllegalConfigurationException;
 import dev.langchain4j.memory.chat.MessageWindowChatMemory;
 import dev.langchain4j.model.chat.ChatLanguageModel;
 import dev.langchain4j.model.chat.mock.ChatModelMock;

--- a/langchain4j/src/test/java/dev/langchain4j/service/AiServiceTokenStreamTest.java
+++ b/langchain4j/src/test/java/dev/langchain4j/service/AiServiceTokenStreamTest.java
@@ -2,7 +2,6 @@ package dev.langchain4j.service;
 
 import dev.langchain4j.data.message.AiMessage;
 import dev.langchain4j.data.message.ChatMessage;
-import dev.langchain4j.exception.IllegalConfigurationException;
 import dev.langchain4j.model.chat.StreamingChatLanguageModel;
 import dev.langchain4j.model.chat.response.ChatResponse;
 import dev.langchain4j.model.output.Response;

--- a/langchain4j/src/test/java/dev/langchain4j/service/AiServicesBuilderTest.java
+++ b/langchain4j/src/test/java/dev/langchain4j/service/AiServicesBuilderTest.java
@@ -2,7 +2,6 @@ package dev.langchain4j.service;
 
 import dev.langchain4j.agent.tool.Tool;
 import dev.langchain4j.data.message.AiMessage;
-import dev.langchain4j.exception.IllegalConfigurationException;
 import dev.langchain4j.model.chat.ChatLanguageModel;
 import dev.langchain4j.model.chat.mock.ChatModelMock;
 import dev.langchain4j.model.output.Response;

--- a/langchain4j/src/test/java/dev/langchain4j/service/AiServicesIT.java
+++ b/langchain4j/src/test/java/dev/langchain4j/service/AiServicesIT.java
@@ -1,6 +1,5 @@
 package dev.langchain4j.service;
 
-import dev.langchain4j.exception.IllegalConfigurationException;
 import dev.langchain4j.memory.ChatMemory;
 import dev.langchain4j.memory.chat.MessageWindowChatMemory;
 import dev.langchain4j.model.chat.ChatLanguageModel;

--- a/langchain4j/src/test/java/dev/langchain4j/service/AiServicesSystemAndUserMessageConfigsTest.java
+++ b/langchain4j/src/test/java/dev/langchain4j/service/AiServicesSystemAndUserMessageConfigsTest.java
@@ -1,6 +1,5 @@
 package dev.langchain4j.service;
 
-import dev.langchain4j.exception.IllegalConfigurationException;
 import dev.langchain4j.model.chat.ChatLanguageModel;
 import dev.langchain4j.model.chat.mock.ChatModelMock;
 import dev.langchain4j.model.chat.request.ChatRequest;

--- a/langchain4j/src/test/java/dev/langchain4j/service/AiServicesUserMessageConfigTest.java
+++ b/langchain4j/src/test/java/dev/langchain4j/service/AiServicesUserMessageConfigTest.java
@@ -1,9 +1,7 @@
 package dev.langchain4j.service;
 
-import dev.langchain4j.exception.IllegalConfigurationException;
 import dev.langchain4j.model.chat.ChatLanguageModel;
 import dev.langchain4j.model.chat.mock.ChatModelMock;
-import dev.langchain4j.model.chat.request.ChatRequest;
 import org.junit.jupiter.api.AfterEach;
 import org.junit.jupiter.api.Test;
 import org.junit.jupiter.api.extension.ExtendWith;
@@ -15,7 +13,6 @@ import static dev.langchain4j.service.AiServicesIT.verifyNoMoreInteractionsFor;
 import static org.assertj.core.api.Assertions.assertThat;
 import static org.assertj.core.api.Assertions.assertThatThrownBy;
 import static org.mockito.Mockito.verify;
-import static org.mockito.Mockito.verifyNoMoreInteractions;
 
 @ExtendWith(MockitoExtension.class)
 class AiServicesUserMessageConfigTest {

--- a/langchain4j/src/test/java/dev/langchain4j/service/output/ServiceOutputParserTest.java
+++ b/langchain4j/src/test/java/dev/langchain4j/service/output/ServiceOutputParserTest.java
@@ -3,7 +3,7 @@ package dev.langchain4j.service.output;
 import com.google.gson.JsonSyntaxException;
 import com.google.gson.reflect.TypeToken;
 import dev.langchain4j.data.message.AiMessage;
-import dev.langchain4j.exception.IllegalConfigurationException;
+import dev.langchain4j.service.IllegalConfigurationException;
 import dev.langchain4j.model.output.Response;
 import dev.langchain4j.model.output.structured.Description;
 import org.junit.jupiter.api.Test;

